### ISSUE TITLE
pylint 1.7.2 RecursionError with luigi

### ISSUE THREAD

**Issue by starhel** [PyCQA/pylint#1642]
### Steps to reproduce
```python
# file: abc.py
import luigi

class SampleReport(luigi.Task):
    def output(self):
        return []

    def run(self):
        for i in luigi.task.flatten(self.output()):
            i.makedirs()
```
```bash
> pylint abc.py
```

### Current behavior
```python
RecursionError: maximum recursion depth exceeded
```
### Expected behavior
Of course I need this working. There's a problem with luigi.task.flatten(). Without this everything is running ok. 

### pylint --version output
```No config file found, using default configuration
pylint 1.7.2, 
astroid 1.5.3
Python 3.5.3 |Continuum Analytics, Inc.| (default, Mar  6 2017, 11:58:13) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]```

**Comment by starhel** [comment]
It's working fine using:
```
pylint 1.6.4, 
astroid 1.4.9
Python 3.5.3 |Continuum Analytics, Inc.| (default, Mar  6 2017, 11:58:13) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]
```

**Comment by brycepg** [comment]
I can reproduce this with 

```python
def func(flat):
    # Needs to be an assignment with modification
    # Needs recursion
    # Cannot be function argument
    flat += func({})
    # Need return, needs to be function argument
    return flat

# Requires assignment, func call
a = func({})
```

**Comment by brycepg** [comment]
Thanks for the report, especially with a good/bad version since I can use git bisect :)

I can repro this without luigi using the flatten function from luigi (see below).

This recursion error is only triggering for me when there is:
- recursion (`flatten` calling itself)
- assignment with modification operators (`+=`)
- calling a method on the result `.out()`/`.pop()`

```python
import six
def flatten(struct):
    """
    Creates a flat list of all all items in structured output (dicts, lists, items):

    .. code-block:: python

        >>> sorted(flatten({'a': 'foo', 'b': 'bar'}))
        ['bar', 'foo']
        >>> sorted(flatten(['foo', ['bar', 'troll']]))
        ['bar', 'foo', 'troll']
        >>> flatten('foo')
        ['foo']
        >>> flatten(42)
        [42]
    """
    flat = []
    if isinstance(struct, dict):
        for _, result in six.iteritems(struct):
            flat += flatten(result)
        return flat
    if isinstance(struct, six.string_types):
        return [struct]
    try:
        # if iterable
        iterator = iter(struct)
    except TypeError:
        return [struct]
    for result in iterator:
        # This needs to be +=
        # Cannot repro with
        # flat = flat + flatten(result)
        flat += flatten(result)
    return flat

flatten([[1], 1])
# Need do so some kind of operation with the result
print(a.pop())
```

**Comment by brycepg** [comment]
This issue is pretty similar to PyCQA/astroid#437
